src/canister/individual_user_template/src/api/cdao/token.rs

```diff
@@ -0,0 +1,79 @@
+use candid::{Nat, Principal};
+use ic_cdk::{query, update};
+use ic_sns_root::pb::v1::{ListSnsCanistersRequest, ListSnsCanistersResponse};
+use icrc_ledger_types::icrc1::{account::Account, transfer::{Memo, TransferArg, TransferError}};
+use shared_utils::{canister_specific::individual_user_template::types::error::CdaoTokenError, pagination::{self, PaginationError}};
+
+use crate::CANISTER_DATA;
+
+/// Add a new token
+/// returns true if new token is added
+#[update]
+async fn add_token(root_canister: Principal) -> Result<bool, CdaoTokenError> {
+    let token_added = CANISTER_DATA.with(|cdata| {
+        let cdata = cdata.borrow();
+        cdata.token_roots.contains_key(&root_canister)
+    });
+    if token_added {
+        return Ok(false);
+    }
+
+    let res: (ListSnsCanistersResponse,) = ic_cdk::call(root_canister, "list_sns_canisters", (ListSnsCanistersRequest {},)).await?;
+    let cans = res.0;
+    let ledger = cans.ledger.ok_or(CdaoTokenError::InvalidRoot)?;
+
+    let acc = Account { owner: ic_cdk::id(), subaccount: None };
+    let balance: (Nat,) = ic_cdk::call(ledger.into(), "icrc1_balance_of", (acc,)).await?;
+    if balance.0 == 0u32 {
+        return Err(CdaoTokenError::NoBalance);
+    }
+
+    CANISTER_DATA.with(|cdata| {
+        let mut cdata = cdata.borrow_mut();
+        cdata.token_roots.insert(root_canister, ());
+    });
+
+
+    return Ok(true);
+}
+
+#[update]
```

Comment:
We need to add a guard here so only the canister owner can call

Reply:
adding